airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py

```diff
@@ -327,6 +327,59 @@ def generate(self) -> Iterable[Partition]:
             )
 
 
+class CursorPartitionGenerator(PartitionGenerator):
+    """
+    This class generates partitions using the concurrent cursor and iterates through state slices to generate partitions.
+
+    It is used when synchronizing a stream in incremental or full-refresh mode where state information is maintained
+    across partitions. Each partition represents a subset of the stream's data and is determined by the cursor's state.
+    """
+
+    def __init__(
+        self,
+        stream: Stream,
+        message_repository: MessageRepository,
+        cursor: Cursor,
+        cursor_field: Optional[List[str]],
+    ):
+        """
+        Initialize the CursorPartitionGenerator with a stream, sync mode, and cursor.
+
+        :param stream: The stream to delegate to for partition generation.
+        :param message_repository: The message repository to use to emit non-record messages.
+        :param sync_mode: The synchronization mode.
+        :param cursor: A Cursor object that maintains the state and the cursor field.
+        """
+        self._stream = stream
+        self.message_repository = message_repository
+        self._sync_mode = SyncMode.full_refresh
+        self._cursor = cursor
+        self._cursor_field = cursor_field
+        self._state = self._cursor.state
+
+    def generate(self) -> Iterable[Partition]:
+        """
+        Generate partitions based on the slices in the cursor's state.
+
+        This method iterates through the list of slices found in the cursor's state, and for each slice, it generates
+        a `StreamPartition` object.
+
+        :return: An iterable of StreamPartition objects.
+        """
+        for slice_start, slice_end in self._cursor.generate_slices():
+            cursor_slice = {"start": slice_start, "end": slice_end}
```

Comment:
> And just to confirm, we're using start and end as the default because abstract_stream_state_converter.py specify those as defaults when slice_boundary_fields is not provided?

Yes, we hard code `start` and `end` in `_add_slice_to_state` for case when we have `_slice_boundary_fields` and in `abstract_stream_state_converter`.